### ISSUE TITLE
Incorrect decoding with pretrained model 

### ISSUE THREAD

**Issue by nmfisher** [foamliu/Speech-Transformer#4]
I've followed the instructions to download the aishell dataset from OpenSLR (http://www.openslr.org/resources/33/data_aishell.tgz) and BEST_checkpoint.tar, and extracted/preprocessed all data.

Below is the output of running demo.py - as you can see, all of these are incorrect.

Did I miss a step somewhere? Or is there possibly a version mismatch between the model in BEST_checkpoint.tar and the data_aishell dataset (i.e. the model was trained on an older token dictionary, but the decoder is using the newer)?

```
OUT_LIST: ['<sos>去化摔第哲鹏态接<eos>', '<sos>去化摔第哲鹏多接<eos>', '<sos>去化摔第哲鹏态背<eos>', '<sos>去化摔第哲鹏多背<eos>', '<sos>去化摔第哲措态接<eos>']
GT: 保障公交路权优先<eos>

OUT_LIST: ['<sos>限筹移限筹冀筹尔球菜<eos>', '<sos>限筹移限筹冀继尔球菜<eos>', '<sos>限筹移限筹应筹尔球菜<eos>', '<sos>限筹移限筹应继尔球菜<eos>', '<sos>限筹移限筹哗筹尔球菜<eos>']
GT: 一零月一零日临近中午<eos>

OUT_LIST: ['<sos>商庭轮镯取方科技态定普将素则<eos>', '<sos>商庭轮镯取方科技态定普将素居<eos>', '<sos>商庭轮镯取方科技态定普将讨则<eos>', '<sos>商庭轮镯取方科技态定普将讨居<eos>', '<sos>商庭轮镯取方科技态定普将目则<eos>']
GT: 在大数据上的技术优势无可匹敌<eos>

OUT_LIST: ['<sos>快多度火讯倒管流商秧试方应招拥埋怨<eos>', '<sos>办多度火讯倒管流商秧试方应招拥埋怨<eos>', '<sos>多多度火讯倒管流商秧试方应招拥埋怨<eos>', '<sos>快多报火讯倒管流商秧试方应招拥埋怨<eos>', '<sos>内多度火讯倒管流商秧试方应招拥埋怨<eos>']
GT: 还有记者爆料说居住在纽约的日本人透露<eos>

OUT_LIST: ['<sos>财多最跨牛密培素饱触<eos>', '<sos>财多最跨牛密万素饱触<eos>', '<sos>财多最夜牛密培素饱触<eos>', '<sos>财多最跨牛密船素饱触<eos>', '<sos>财多最跨牛密培素委触<eos>']
GT: 没有发生任何肢体冲突<eos>

OUT_LIST: ['<sos>商泪油义界喊泪方次针项部界些泪方次针项部<eos>', '<sos>商泪油义界喊泪方次针项部界较泪方次针项部<eos>', '<sos>商泪油义界喊泪方次针项部部些泪方次针项部<eos>', '<sos>商泪油义界些泪方次针项部界些泪方次针项部<eos>', '<sos>商泪油义界喊泪方次针项部部较泪方次针项部<eos>']
GT: 在类似印度之类的新兴市场<eos>

OUT_LIST: ['<sos>收内潜殴持市率钉能激酒市率限影酒市率敏酒<eos>', '<sos>收冯潜殴持市率钉能激酒市率限影酒市率敏酒<eos>', '<sos>收内潜苏持市率钉能激酒市率限影酒市率敏酒<eos>', '<sos>收规潜殴持市率钉能激酒市率限影酒市率敏酒<eos>', '<sos>收内潜练持市率钉能激酒市率限影酒市率敏酒<eos>']
GT: 你要懂既然不是电子产品不是易损品不是玩品<eos>

OUT_LIST: ['<sos>军条脸方这喜居取佶有笙栈尔应<eos>', '<sos>军条脸方这喜居取幕有笙栈尔应<eos>', '<sos>军条脸方这杭居取佶有笙栈尔应<eos>', '<sos>军条脸方这杭居取幕有笙栈尔应<eos>', '<sos>军条脸方这喜居取佶贷笙栈尔应<eos>']
GT: 小区外的空置地上填满垃圾近日<eos>

OUT_LIST: ['<sos>依竞厂遵练角最墙浪称球房银津嘉<eos>', '<sos>依竞厂遵练角最墙浪称球房银津嘉嘉<eos>', '<sos>依竞厂究练角最墙浪称球房银津嘉<eos>', '<sos>依竞厂监练角最墙浪称球房银津嘉<eos>', '<sos>依竞厂遵练角最墙浪称球房银津嘉脑<eos>']
GT: 引导金融机构发放农业中长期贷款<eos>

OUT_LIST: ['<sos>多底仍供汗资乓姓<eos>', '<sos>多底仍供赃资乓姓<eos>', '<sos>多底仍供汗炉乓姓<eos>', '<sos>多底仍供赃炉乓姓<eos>', '<sos>多底仍供汗令乓姓<eos>']
GT: 有时甚至忘记吃饭<eos>
```

**Comment by wwwjjjnnn** [comment]
I have the same problem as you.
I think your guess is correct, So I modified demo.py.

`import argparse
import pickle
import random
from shutil import copyfile

import torch
import copy

from config import pickle_file, device, input_dim, LFR_m, LFR_n
from data_gen import build_LFR_features
from utils import extract_feature, ensure_folder
from xer import cer_function


def parse_args():
    parser = argparse.ArgumentParser(
        "End-to-End Automatic Speech Recognition Decoding.")
    # decode
    parser.add_argument('--beam_size', default=5, type=int,
                        help='Beam size')
    parser.add_argument('--nbest', default=5, type=int,
                        help='Nbest size')
    parser.add_argument('--decode_max_len', default=100, type=int,
                        help='Max output length. If ==0 (default), it uses a '
                             'end-detect function to automatically find maximum '
                             'hypothesis lengths')
    args = parser.parse_args()
    return args


if __name__ == '__main__':
    args = parse_args()
    with open(pickle_file, 'rb') as file:
        data = pickle.load(file)
    char_list_new = data['IVOCAB']
    ichar_list_new = data['VOCAB']
    samples = data['test']

    with open('char_list.pkl', 'rb') as file:
        char_list = pickle.load(file)
        ichar_list = dict(zip(char_list.values(), char_list.keys()))

    checkpoint = 'BEST_checkpoint.tar'
    checkpoint = torch.load(checkpoint)
    model = checkpoint['model']
    model.eval()

    samples = random.sample(samples, 10)
    ensure_folder('audios')
    results = []

    for i, sample in enumerate(samples):
        wave = sample['wave']
        trn = sample['trn']

        trn_new = []
        for index in trn:
            trn_new.append(ichar_list[char_list_new[index]])
        trn = copy.deepcopy(trn_new)

        copyfile(wave, 'audios/audio_{}.wav'.format(i))

        feature = extract_feature(input_file=wave, feature='fbank', dim=input_dim, cmvn=True)
        feature = build_LFR_features(feature, m=LFR_m, n=LFR_n)
        # feature = np.expand_dims(feature, axis=0)
        input = torch.from_numpy(feature).to(device)
        input_length = [input[0].shape[0]]
        input_length = torch.LongTensor(input_length).to(device)
        nbest_hyps = model.recognize(input, input_length, char_list, args)
        out_list = []
        for hyp in nbest_hyps:
            out = hyp['yseq']
            out = [char_list[idx] for idx in out]
            out = ''.join(out)
            out_list.append(out)
        print('OUT_LIST: {}'.format(out_list))

        gt = [char_list[idx] for idx in trn]
        gt = ''.join(gt)
        print('GT: {}\n'.format(gt))

        results.append({'out_list_{}'.format(i): out_list, 'gt_{}'.format(i): gt})

    import json

    with open('results.json', 'w') as file:
        json.dump(results, file, indent=4, ensure_ascii=False)
`

**Comment by wjunneng** [comment]
I have the same problem as you.
I think your guess is correct, So I modified demo.py and run success.

![截图](https://user-images.githubusercontent.com/37031645/68753561-3845e600-0640-11ea-8611-710f8871f335.png)

**Comment by foamliu** [comment]
@nmfisher @wwwjjjnnn Thanks for trying. It was my bad: "BEST_checkpoint.tar" is not used any longer, and the right model file is: "speech-transformer-cn.pt". I've updated the source code in demo.py, please try again and let me know if the issue still exists.
